### PR TITLE
Add config option to change the default widget cache time

### LINKED ISSUE #1170: Make the default widget cache time configurable
Currently the default cache time is a constant in https://github.com/Graylog2/graylog2-server/blob/master/graylog2-server/src/main/java/org/graylog2/dashboards/widgets/DashboardWidget.java#L63

For very complex dashboards in high traffic setups, it can be beneficial for the operator to set a more reasonable default timeout to prevent accidental generation of many very short lived widget cache times.

Introduce a server.conf setting to override the default time.

### PR DESCRIPTION
**Pull request by bernd**
This is my 4th try to make the default widget cache time configurable and it is still a huge change. :confused: I think this version is a step in the right direction to make the dashboard widget code better in the future. (removal of static methods, dependency injection)

Please let me know if you can think of a better way that does not introduce such big changes!

* Refactors static DashboardWidget creator methods into DashboardWidgetCreator and replace usage of static methods with creator object which gets injected.
* Removes the hardcoded default dashboard widget cache time and makes it configurable with the "dashboard_widget_default_cache_time" config file option.
* Introduce a WidgetCacheTime object that decides when the default cache time is used.

Fixes #1170

### CODE REVIEW

File: graylog2-server/src/main/java/org/graylog2/Configuration.java [hunk 1]
```diff
@@ -130,6 +130,9 @@
     @Parameter(value = "collector_inactive_threshold", validator = PositiveIntegerValidator.class)
     private Duration collectorInactiveThreshold = Duration.minutes(1);
 
+    @Parameter(value = "dashboard_widget_default_cache_time", validator = PositiveIntegerValidator.class)
+    private int dashboardWidgetDefaultCacheTime = 10;
```
Comment (joschi) [inline graylog2-server/src/main/java/org/graylog2/Configuration.java]: Why not use a [`Duration`](https://joschi.github.io/JadConfig/0.10.1/apidocs/com/github/joschi/jadconfig/util/Duration.html) (or the equivalent from JodaTime) here?

File: graylog2-server/src/main/java/org/graylog2/dashboards/widgets/DashboardWidgetCreator.java [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+import org.joda.time.DateTimeZone;
+
+import javax.inject.Inject;
+import java.util.Map;
+import java.util.UUID;
+
+import static com.google.common.base.MoreObjects.firstNonNull;
+import static com.google.common.base.Strings.isNullOrEmpty;
+
+public class DashboardWidgetCreator {
```
Comment (joschi) [inline graylog2-server/src/main/java/org/graylog2/dashboards/widgets/DashboardWidgetCreator.java]: I'd have called it `DashboardWidgetFactory`. :wink:

File: graylog2-server/src/main/java/org/graylog2/dashboards/widgets/WidgetCacheTime.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    private final int cacheTime;
+
+    public interface Factory {
+        WidgetCacheTime create(int cacheTime);
+    }
+
+    @Inject
+    public WidgetCacheTime(@Named("dashboard_widget_default_cache_time") int defaultCacheTime,
+                           @Assisted int cacheTime) {
+        this.cacheTime = cacheTime < 1 ? defaultCacheTime : cacheTime;
```
Comment (joschi) [inline graylog2-server/src/main/java/org/graylog2/dashboards/widgets/WidgetCacheTime.java]: I know that this was the standard behavior before, but shouldn't we allow a cache time of 0 to deactivate caching at all (if people really wanted that)?

Comment (kroepke) [inline graylog2-server/src/main/java/org/graylog2/dashboards/widgets/WidgetCacheTime.java]: I think we should keep it, as a short caching time also has the effect of limiting the impact multiple users/clients have: All of their requests are fast, and do not increase the search load.

Also if we consider that most deployments will run with a ES refresh time of 30 seconds or more, then turning off the cache will have no beneficial impact.

File: graylog2-server/src/main/java/org/graylog2/Configuration.java (hunk 1 above)
Comment (bernd) [inline graylog2-server/src/main/java/org/graylog2/Configuration.java]: Good point, will change that!

File: graylog2-server/src/main/java/org/graylog2/dashboards/widgets/DashboardWidgetCreator.java (hunk 2 above)
Comment (bernd) [inline graylog2-server/src/main/java/org/graylog2/dashboards/widgets/DashboardWidgetCreator.java]: Factory sounds like hard work, creator sounds like an artist creating beautiful dashboard widgets. :stuck_out_tongue:

**Comment by bernd** [comment]
Updated with `Duration` usage. I would also keep the minimum cache time of 1 second for now.

**Comment by bernd** [comment]
Also updated the example config file.

File: graylog2-server/src/main/java/org/graylog2/dashboards/widgets/DashboardWidgetCreator.java (hunk 2 above)
Comment (joschi) [inline graylog2-server/src/main/java/org/graylog2/dashboards/widgets/DashboardWidgetCreator.java]: We're all beautiful unicorns!
Unicorn

**Comment by joschi** [comment]
LGTM (except for this `DashboardWidgetCreator` thing :stuck_out_tongue:). :+1:

**Pull request merged**